pkg/controller/infrastructure/infraflow/ensurer.go

```diff
@@ -699,6 +742,88 @@ func (fctx *FlowContext) EnsureManagedIdentity(ctx context.Context) (err error)
 	return err
 }
 
+// MigrateAvailabilitySet prepares an AS-based shoot to be migrated to VMSS-Flex.
+func (fctx *FlowContext) MigrateAvailabilitySet(ctx context.Context) error {
+	var (
+		log = shared.LogFromContext(ctx)
+		c   = fctx.client
+	)
+
+	// return early if the migration has already been complete
+	if v := fctx.whiteboard.GetChild("migration").GetChild(KindAvailabilitySet.String()).Get("complete"); v != nil && *v == "true" {
+		return nil
+	}
+	// return early if the cluster does not have AS.
+	if fctx.whiteboard.GetChild(ChildKeyIDs).Get(KindAvailabilitySet.String()) == nil {
+		return nil
+	}
+
+	// IF VMOs are not needed, or the migration is already done, return early.
+	if !helper.HasShootVmoMigrationAnnotation(fctx.cluster.Shoot.GetAnnotations()) {
+		return nil
+	}
```

Comment:
These checks as a group are used in one place. The separate function will just be giving a name to preflight checks but I can oblige

Reply:
- Do you mean as a `shared.DoIf` ?
It's a question of when are the dependencies of a task evaluated (at build or at runtime) and AFAIK they are on build time. At that point you may not have the `KindAvailabilitySet` in the state for example. 

- Do you mean as a `shared.Dependency` ? 
I am actually not sure if this will block a child Task in the graph from being executed. I would need to test the scemantics. It may either mark the skipped parent as completed (so the childen are always executed), or it may really prune the tree (what you suggested I believe). But I need to check first.

Reply:
Based on https://github.com/gardener/gardener/blob/84184385dbb9616c2c838e4f24487cae2be75da0/pkg/utils/flow/flow.go#L303 I think it marks the skipped nodes as completed

Reply:
Valid arguments.
I think I was a bit too quick here and didn't think it through completely.
Let's leave it as is.